File: go/api/jdy/form_data.go
/**
 * Copyright (c) 2015-2022, FineX, All Rights Reserved.
 * @author Thomas.Zhuang
 * @date 2022/5/17
 */

package jdy

import (
	"api-demo/base"
	"api-demo/model"
	"encoding/json"
	"fmt"
)

// FormDataApiClient 表单数据Api
type FormDataApiClient struct {
	*base.ApiClient
}

func (api *FormDataApiClient) ValidVersion(version string) bool {
	return map[string]bool{
		"v1": true,
		"v2": true,
		"v3": true,
		"v4": true,
	}[version]
}

func (api *FormDataApiClient) DefaultVersion() string {
	return "v4"
}

// DoRequest 发送http请求
func (api *FormDataApiClient) DoRequest(option *base.RequestOption) (responseBody []byte, err error) {
	if !api.ValidVersion(api.Version) {
		api.Version = api.DefaultVersion()
	}
	return api.ApiClient.DoRequest(option)
}

// SingleDataCreate 新建单条数据接口
func (api *FormDataApiClient) SingleDataCreate(appId, entryId string, data, option map[string]interface{}) (responseBody *model.FormData, err error) {
	option["data"] = data
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_create", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// SingleDataQuery 查询单条数据接口
func (api *FormDataApiClient) SingleDataQuery(appId, entryId, dataId string) (responseBody *model.FormData, err error) {
	payload, err := json.Marshal(map[string]interface{}{
		"data_id": dataId,
	})
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_retrieve", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// SingleDataUpdate 修改单条数据接口
func (api *FormDataApiClient) SingleDataUpdate(appId, entryId, dataId string, data, option map[string]interface{}) (responseBody *model.FormData, err error) {
	option["data"] = data
	option["data_id"] = dataId
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_update", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// SingleDataRemove 删除单条数据接口
func (api *FormDataApiClient) SingleDataRemove(appId, entryId, dataId string, option map[string]interface{}) (responseBody *model.Response, err error) {
	option["data_id"] = dataId
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Version: "v1",
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_delete", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// BatchDataCreate 新建多条数据接口
func (api *FormDataApiClient) BatchDataCreate(appId, entryId string, dataList []map[string]interface{}, option map[string]interface{}) (responseBody *model.BatchResponse, err error) {
	option["data_list"] = dataList
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Version: "v1",
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_batch_create", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// BatchDataQuery 查询多条数据接口
func (api *FormDataApiClient) BatchDataQuery(appId, entryId string, option map[string]interface{}) (responseBody *model.FormDataList, err error) {
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Version: "v1",
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// BatchDataUpdate 修改多条数据接口
func (api *FormDataApiClient) BatchDataUpdate(appId, entryId string, dataIds []string, data map[string]interface{}, option map[string]interface{}) (responseBody *model.BatchResponse, err error) {
	option["data_ids"] = dataIds
	option["data"] = data
	payload, err := json.Marshal(option)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Version: "v1",
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_batch_update", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// BatchDataRemove 删除多条数据接口
func (api *FormDataApiClient) BatchDataRemove(appId, entryId string, dataIds []string) (responseBody *model.BatchResponse, err error) {
	payload, err := json.Marshal(map[string]interface{}{
		"data_ids": dataIds,
	})
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Version: "v1",
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/data_batch_delete", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

File: csharp/src/api/jdy/FormData.cs
/**
* Copyright (c) 2015-2022, FineX, All Rights Reserved.
* @author Thomas.Zhuang
* @date 2022/06/14
*/
using System.Text.Json;
using Src.Base;

namespace Src.Api.Jdy;

public class FormDataApiClient : ApiClient
{
    private readonly static string[] validVersions = { "v4", "v3", "v2", "v1" };
    private readonly static string defaultVersion = "v4";
    public FormDataApiClient(string host, string apiKey, string version = "") : base(host, apiKey, version)
    {
    }

    public new async Task<JsonElement?> doRequest(IDictionary<string, object> option)
    {
        if (Array.IndexOf(validVersions, base.version) == -1)
        {
            base.version = defaultVersion;
        }
        return await base.doRequest(option);
    }

    /**
    * 新建单条数据接口
    */
    public async Task<JsonElement?> singleDataCreate(string appId, string entryId, IDictionary<string, object> data, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_create");
        option.Add("data", data);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 查询单条数据接口
    */
    public async Task<JsonElement?> singleDataQuery(string appId, string entryId, string dataId)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_retrieve");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("data_id", dataId);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }

    /**
    * 修改单条数据接口
    */
    public async Task<JsonElement?> singleDataUpdate(string appId, string entryId, string dataId, IDictionary<string, object> data, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_update");
        option.Add("data", data);
        option.Add("data_id", dataId);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 删除单条数据接口
    */
    public async Task<JsonElement?> singleDataRemove(string appId, string entryId, string dataId, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("version", "v1");
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_delete");
        option.Add("data_id", dataId);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 新建多条数据接口
    */
    public async Task<JsonElement?> batchDataCreate(string appId, string entryId, IDictionary<string, object>[] dataList, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("version", "v1");
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_batch_create");
        option.Add("data_list", dataList);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 查询多条数据接口
    */
    public async Task<JsonElement?> batchDataQuery(string appId, string entryId, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("version", "v1");
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data");
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 修改多条数据接口
    */
    public async Task<JsonElement?> batchDataUpdate(string appId, string entryId, string[] dataIds, IDictionary<string, object> data, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("version", "v1");
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_batch_update");
        option.Add("data_ids", dataIds);
        option.Add("data", data);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 删除多条数据接口
    */
    public async Task<JsonElement?> batchDataRemove(string appId, string entryId, string[] dataIds)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("version", "v1");
        requestOption.Add("method", "POST");
        requestOption.Add("path", $"app/{appId}/entry/{entryId}/data_batch_delete");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("data_ids", dataIds);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }
}